src/services/trackingPlan.ts

```diff
@@ -1,88 +1,83 @@
-            statusCode: 200,
-            validationErrors: hv['validationErrors'],
-            error: JSON.stringify(constructValidationErrors(hv['validationErrors'])),
-          });
-          stats.counter('tp_propagated_events', 1, {
-            ...metaTags,
-          });
-        }
+        event.request = { query: reqParams };
+        const validatedEvent = await eventValidator.handleValidation(event);
+        eventValidationResponse = {
+          output: event.message,
+          metadata: event.metadata,
+          statusCode: validatedEvent['dropEvent']
+            ? HTTP_STATUS_CODES.BAD_REQUEST
+            : HTTP_STATUS_CODES.OK,
+          validationErrors: validatedEvent['validationErrors'],
+          error: JSON.stringify(constructValidationErrors(validatedEvent['validationErrors'])),
+        };
       } catch (error) {
-        const errMessage = `Error occurred while validating : ${error}`;
-        logger.error(errMessage);
-        let status = 200;
+        logger.debug(
+          `Error occurred while validating event`,
+          'event',
+          event?.eventName,
+          'tp',
+          event?.trackingPlanId,
+          'tpVersion',
+          event?.trackingPlanVersion,
+          'error',
+          error,
+        );
+
+        exceptionOccured = true;
+        // no need to process further if
+        // we have error of retry request error
         if (error instanceof RetryRequestError) {
           ctxStatusCode = error.statusCode;
+          break;
         }
-        if (error instanceof RespStatusError) {
-          status = error.statusCode;
-        }
-        respList.push({
+
+        eventValidationResponse = {
           output: event.message,
           metadata: event.metadata,
-          statusCode: status,
+          statusCode: error instanceof RespStatusError ? error.statusCode : HTTP_STATUS_CODES.OK,
           validationErrors: [],
-          error: errMessage,
-        });
-        stats.counter('tp_errors', 1, {
-          ...metaTags,
-          workspaceId: event.metadata?.workspaceId,
-          trackingPlanId: event.metadata?.trackingPlanId,
-        });
+          error: `Error occurred while validating: ${error}`,
+        };
       } finally {
-        stats.timing('tp_event_latency', eventStartTime, {
+        // finally on every event, we need to
+        // capture the information related to the validates event
+        stats.timing('tp_event_validation_latency', eventStartTime, {
           ...metaTags,
+          ...tpTags,
+          status: eventValidationResponse.statusCode,
+          exception: exceptionOccured,
         });
       }
-    }
-
-    stats.counter('tp_events_count', events.length, {
-      ...metaTags,
-    });
 
-    stats.histogram('tp_request_size', requestSize, {
-      ...metaTags,
-    });
+      respList.push(eventValidationResponse);
+    }
 
-    stats.timing('tp_request_latency', requestStartTime, {
+    // capture overall function latency
```

Comment:
don't remove `stats.histogram('tp_request_size', requestSize)`. It is a useful metric too